bin/vim_erlang_tags.erl

```diff
@@ -120,77 +120,126 @@ get_command_type(C) when C=:=include; C=:=ignore; C=:=output -> stateful;
+         | {cmd_line_arguments(), cmd_line_arguments()}.
+get_full_arg_state(Param, _CurrentParamState, ToContinueParsing)
+  when Param =:= otp; Param =:= help; Param =:= verbose ->
+    {true, ToContinueParsing};
+get_full_arg_state(Param, CurrentParamState, ToContinueParsing) ->
+    log("Parsing args for parameter ~p~n", [Param]),
+    {StateArgs, Rest} = consume_until_new_state(ToContinueParsing),
+    case StateArgs of
+        [] -> log_error("Arguments needed for ~s.~n", [Param]);
+        _ -> ok
     end,
-    put(ignored, AllIgnored),
-    parse_args(OtherArgs);
-parse_args(["-" ++ Arg|_]) ->
-    log_error("Unknown argument: ~s~n", [Arg]),
-    halt(1);
-parse_args([FileName|OtherArgs]) ->
-    put(files, [FileName|get(files)]),
-    parse_args(OtherArgs).
+    {StateArgs ++ CurrentParamState, Rest}.
+
+-spec consume_until_new_state(Args) -> {ConsumedArgs, RestArgs} when
+      Args :: cmd_line_arguments(),
+      ConsumedArgs :: cmd_line_arguments(),
+      RestArgs :: cmd_line_arguments().
+consume_until_new_state(Args) ->
+    log("    Consuming args ~p~n", [Args]),
+    States = lists:foldl(fun({_,S}, Acc) -> S ++ Acc end, [], allowed_cmd_params()),
+    lists:splitwith(
+      fun("-" ++ _ = El) ->
+              case lists:member(El, States) of
+                  true -> false;
+                  _ -> log_error("Unknown argument: ~s~n", [El]), halt(1)
+              end;
+         (El) ->
+              not lists:member(El, States)
+      end, Args).
+
+-spec clean_opts(parsed_params()) -> config().
+clean_opts(#{help := true}) ->
+    #{help => true};
+clean_opts(#{include := []} = Opts0) ->
+    log("Set includes to default current dir.~n"),
+    clean_opts(Opts0#{include := [?DEFAULT_PATH]});
+clean_opts(#{otp := true, include := Inc} = Opts0) ->
+    log("Including OTP in.~n"),
+    AllIncludes = [code:lib_dir() | Inc],
+    Opts1 = maps:update(include, AllIncludes, Opts0),
+    Opts2 = maps:update(otp, false, Opts1),
+    clean_opts(Opts2);
+clean_opts(#{output := []} = Opts0) ->
+    log("Set output to default 'tags'.~n"),
+    clean_opts(Opts0#{output := ["tags"]});
+clean_opts(#{include := Included, ignore := Ignored, output := [Output]}) ->
+    log("Set includes to default current dir.~n"),
+    #{explore => expand_includes_remove_ignored(Included, Ignored), output => Output}.
+
+-spec expand_includes_remove_ignored([string()], [string()]) -> [file:filename()].
+expand_includes_remove_ignored(Included, Ignored) ->
+    AllIncluded = lists:foldl(fun(L,Acc) -> L++Acc end, [], expand_dirs(Included)),
+    AllIgnored = lists:foldl(fun(L,Acc) -> L++Acc end, [], expand_dirs(Ignored)),
+    lists:subtract(AllIncluded, AllIgnored).
+
+-spec expand_dirs([string()]) -> [file:filename()].
+expand_dirs(Included) ->
+    lists:map(
+      fun(FileName) ->
+              case filelib:is_file(FileName) of
+                  false ->
+                          log_error("FileName: ~p~n", [FileName]),
```

Comment:
1. I suggest a more descriptive error message.
2. The indentation is too much.

Reply:
Indeed, this one now goes as
```erlang
-spec expand_dirs([string()]) -> [file:filename()].
expand_dirs(Included) ->
    lists:map(fun expand_dirs_or_filenames/1, Included).

-spec expand_dirs_or_filenames(string()) -> [file:filename()].
expand_dirs_or_filenames(FileName) ->
    case {filelib:is_file(FileName), filelib:is_dir(FileName)} of
        {false, _} ->
            log_error("File \"~p\" is not a proper file.~n", [FileName]),
            [];
        {true, true} ->
                    filelib:wildcard(FileName ++ "/**/*.{erl,hrl}");
        _ -> [FileName]
    end.
```